Add AckDeadlineRenewer class for automatic ack deadline renewal

mziccard opened the pull request:
This is a simpler version of #1024 to provide support for automatic ack deadline renewals

mziccard commented on gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSubOptions.java:
```diff
@@ -86,6 +86,11 @@ protected PubSubOptions(Builder builder) {
   }
 
   @Override
+  protected ExecutorFactory executorFactory() {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Overriding method merely calls super](https://www.codacy.com/app/mziccard/gcloud-java/file/2647881484/issues/source?bid=3123109&fileBranchId=3323306#l89)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Overriding method merely calls super](https://www.codacy.com/app/mziccard/gcloud-java/file/2648334987/issues/source?bid=3123109&fileBranchId=3323306#l89)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Overriding method merely calls super](https://www.codacy.com/app/mziccard/gcloud-java/file/2648419172/issues/source?bid=3123109&fileBranchId=3323306#l89)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Overriding method merely calls super](https://www.codacy.com/app/mziccard/gcloud-java/file/2648607896/issues/source?bid=3123109&fileBranchId=3323306#l89)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Overriding method merely calls super](https://www.codacy.com/app/mziccard/gcloud-java/file/2670963122/issues/source?bid=3123109&fileBranchId=3323306#l89)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Overriding method merely calls super](https://www.codacy.com/app/mziccard/gcloud-java/file/2673950796/issues/source?bid=3123109&fileBranchId=3323306#l89)

mziccard replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Overriding method merely calls super](https://www.codacy.com/app/mziccard/gcloud-java/file/2681421883/issues/source?bid=3123109&fileBranchId=3323306#l89)

lesv replied:
So why have it?

mziccard replied:
So that the method is visible in the package.

mziccard commented:
/cc @aozarov

aozarov commented on gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/AckDeadlineRenewer.java:
```diff
@@ -0,0 +1,268 @@
+ * {@link #add(String, Iterable)}. The acknowledge deadlines of added messages are renewed until the
+ * messages are explicitly removed using {@link #remove(String, String)}.
+ */
+class AckDeadlineRenewer implements AutoCloseable {
```
Overall implementation looks fine.

I think it can be improved (e.g. wake up as needed and not based on a constant internal, auto-remove items that are known to have expired, handle deadline renewal failures, minimize possible RPC by including more ackIds for a subscription if we are already going to make an RPC for that subscription,...) however, starting with simple and improving, if needed, later sounds like a right move.

aozarov commented on gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/AckDeadlineRenewer.java:
```diff
@@ -0,0 +1,268 @@
+ */
+class AckDeadlineRenewer implements AutoCloseable {
+
+  private static final int MIN_DEADLINE_MILLISECONDS = 10_000;
```
maybe for defaults we should have a bit more slack in both the min and the threshold. I would consult the PubSub team for that.

aozarov commented on gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/AckDeadlineRenewer.java:
```diff
@@ -0,0 +1,268 @@
+
+  @Override
+  public void close() throws Exception {
+    renewerFuture.cancel(false);
```
this could have clear messageQueue and/or messageDeadlines.

mziccard replied:
Agree, I was also planning to change to `cancel(true)` and add a `closed` field to avoid doing the cleanup twice.

aozarov commented on gcloud-java-pubsub/src/test/java/com/google/cloud/pubsub/AckDeadlineRenewerTest.java:
```diff
@@ -0,0 +1,234 @@
+
+  @Before
+  public void setUp() {
+    clock = EasyMock.createStrictMock(Clock.class);
```
This test seems to me too fragile/internal. I am not sure why we could not use non mocked executor and use either Clock or enable setting the deadline (package scope, testing only), so only mock and verification would be PubSub.

mziccard replied:
I agree that this test is very internal but I don't see this as necessarily bad. How would using actual executors help? If we mock clock and use actual executors we will end up with very similar tests plus the variability given by the fact that we are actually using threads. I am not sure it's a good idea in general to use executors for unit testing.

aozarov replied:
The more you mock the more susceptible you are to implementation details changes and in some cases you may over-mock and ignore potential issues.

I think it is reasonable to have some tests with timeouts and  you can use the `PusSub` last expected call as away indicate completion  (e.g. look at the first answer [here](http://stackoverflow.com/questions/16651578/using-easymock-in-multithreading)).

Just something to consider...

mziccard replied:
> Just something to consider...

I'll give this a try tomorrow and see if I can reduce mocking while still providing adequate coverage. My fear is that if with using actual executors we will end up testing the same behavior as before but also add the risk of tests being flaky. But again, let me give this a try :)

mziccard commented:
@aozarov In d4b8f6c I refactored the tests to follow your suggestion. I also addressed other comments. I removed the fixed interval and now renewer is only scheduled when actually needed. PTAL

aozarov commented on gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/AckDeadlineRenewer.java:
```diff
@@ -0,0 +1,316 @@
+    }
+    for (Map.Entry<String, List<String>> entry : Multimaps.asMap(messagesToRenewNext).entrySet()) {
+      // We send all ack deadline renewals for a subscription
+      pubsub.modifyAckDeadlineAsync(entry.getKey(), MIN_DEADLINE_MILLIS,
```
I actually think that we should ask to lease a bit *longer* than the 10 seconds minimum however assume that we only got 10 seconds and use slack before we renew again.

aozarov commented:
Change looks good. Implementation looks good and api looks good (I provided one more minor suggestion). Maybe @lesv want to go over this code as well (I didn't pay attention to anything but API and impl).

mziccard commented:
I am merging this, if @lesv has some more comments I'd be happy to address them in separate PRs.

lesv commented:
I'm happy with this change.
